File: wakeup.pl
#!/usr/bin/perl

use strict;
use warnings;
use LWP::UserAgent;
use Encode;
use Digest::MD5 qw(md5_hex);
use JSON;

my $ua = LWP::UserAgent->new;

### start of config section

# if you don't want to store your credentials within this script just
# keep the next three dummy values and set $ask_auth = 1

# the following credentials are needed for remote login, e.g. from WAN/Internet
my $remote_user = "myUsername";
my $remote_pass = "myPassword";

# local admin password required for some fritzboxes 
my $local_admin_pass = "myAdminPassword";

# enable interactive authentication to avoid storing credentials in this script
my $ask_auth = 0;

# fritzbox host to connect to
my $host = "my.hostname.net";

# fritzbox port, usually 443
my $port = 443;

# MAC of the computer to send the wakeup packet to,
# use mac format 01:23:45:67:89:AB
my $mac = "00:CO:1D:CO:FF:EE";

# uncomment the following line to disable ssl certificate verification
#$ua->ssl_opts(verify_hostname => 0, SSL_verify_mode => 0);

### end of config section
### do not change anything below!

sub err_exit($) {
    my $msg = shift;
    print "[error] $msg\n";
    exit 1;
}

sub readcreds() {
    use Term::ReadKey;
    print "remote login username: "; chomp($remote_user = ReadLine(0));
    ReadMode('noecho');
    print "remote login password: "; chomp($remote_pass = ReadLine(0));
    print "\nlocal admin password: "; chomp($local_admin_pass = ReadLine(0));
    ReadMode('restore');
    print "\n\n";
}

readcreds if ($ask_auth);

$ua->timeout(30);
push @{$ua->requests_redirectable}, 'POST';
$ua->credentials("$host:$port", "HTTPS Access", $remote_user, $remote_pass);

my $url_base = "https://$host:$port";
my $url_login_suffix = "/login.lua";
my $url_login = $url_base.$url_login_suffix;
my $url_netdevs = "$url_base/net/network_user_devices.lua?sid=";
my $url_editdev_suffix = "net/edit_device.lua";
my $url_wakeup = "$url_base/$url_editdev_suffix?sid=";
my $url_wakeup_old = "$url_base/cgi-bin/webcm";
my $url_data = "$url_base/data.lua";

my $r = $ua->get($url_login);
if ($r->is_success) {
    my $c = $r->decoded_content;
    if ($c =~ m#(?:g_challenge|"challenge"|var challenge|\["security:status/challenge"\])(?: =|:) "([a-f0-9]+)"#) {
        my $challenge = $1;
        my $pass = ($c =~ m#Benutzername#) ? $remote_pass : $local_admin_pass;
        my %data = (
            username => $remote_user,
            response => sprintf "%s-%s", $challenge, md5_hex(encode("UTF16-LE", sprintf "%s-%s", $challenge, $pass))
        );
        my $vers = 0;
        my $target = ($c =~ m#action="$url_login_suffix"#) ? $url_login : $url_base;
        $r = $ua->post($target, \%data);
        $c = $r->decoded_content;
        if ($c =~ m#(?:FRITZ!OS |version%3D\d\d\.)(\d+)\.(\d+)#) {
            $vers = sprintf "%d", $1.$2;
        }
        err_exit "login failed" if ($c =~ m#(?:error_text|ErrorMsg)#i);
        if ($c =~ m#(?:(?:home|logout)\.lua\?sid=|"sid": ")([a-f0-9]+)#) {
            my $sid = $1;
            if ($vers < 630) {
                %data = (sid => $sid, "wakeup:settings/mac" => $mac);
                $r = $ua->post($url_wakeup_old, \%data);
                if ($r->is_success) {
                    print "[success] wakeup done\n";
                } else {
                    err_exit "invalid post data" . $r->status_line;
                }
            } elsif ($vers < 650) {
                $r = $ua->get($url_netdevs.$sid);
                $c = $r->decoded_content;
                if ($c =~ m#$mac.*? value="([^"]+)"#i) {
                    my $uid = $1;
                    %data = (dev => $uid, btn_wake => "");
                    $r = $ua->post($url_wakeup.$sid, \%data);
                    if ($r->is_success) {
                        print "[success] wakeup done\n";
                    } else {
                        err_exit "invalid post data: " . $r->status_line;
                    }
                } else {
                    err_exit "could not find mac";
                }
            } else {
                %data = (
                    sid => $sid,
                    page => "netDev"
                );
                $r = $ua->post($url_data, \%data);
                $c = $r->decoded_content;
                my $json = from_json($c);
                my $uid = "";
                my @devs = @{$json->{'data'}->{'passive'}};
                push @devs, @{$json->{'data'}->{'active'}};
                foreach my $dev (@devs) {
                    if ($dev->{'mac'} eq $mac) {
                        $uid = $dev->{'UID'};
                        last;
                    }
                }
                if ($uid ne "") {
                    %data = (
                        sid => $sid,
                        dev => $uid,
                        oldpage => $url_editdev_suffix,
                        btn_wake => ""
                    );
                    $r = $ua->post($url_data, \%data);
                    $c = $r->decoded_content;
                    if ($c =~ m#"pid": "netDev"#) {
                        print "[success] wakeup done\n";
                    } else {
                        err_exit "invalid post data: " . $r->status_line;
                    }
                } else {
                    err_exit "could not find mac"
                }
            }
        } else {
            err_exit "could not find a session id";
        }
    } else {
        err_exit "could not find a challenge";
    }
} else {
    err_exit "could not load login page: " . $r->status_line;
}

exit 0;

File: wakeup.py
#!/usr/bin/env python
# coding: utf-8
#
# (c)2017-2021 n0rc

import argparse
import getpass
import hashlib
import json
import os
import requests
import sys

from lxml import etree
from packaging import version
from requests.exceptions import SSLError
from requests.packages.urllib3.exceptions import InsecureRequestWarning

requests.packages.urllib3.disable_warnings(InsecureRequestWarning)

SID_NOAUTH = '0000000000000000'


def error_exit(msg):
    print("[error] {}".format(msg))
    sys.exit(1)


def ssl_error_exit():
    error_exit("ssl certificate verification failed")


def get_config(config_file):
    if os.path.isfile(config_file):
        with open(config_file, 'r') as jf:
            config = json.load(jf)
        config['url_login'] = 'https://{}:{}/login_sid.lua'.format(config['host'], config['port'])
        config['url_data'] = 'https://{}:{}/data.lua'.format(config['host'], config['port'])
        return config
    else:
        error_exit("config file not found")


def get_sid(config):
    try:
        password = config['password']
    except:
        password = getpass.getpass()
    try:
        r = requests.get(config['url_login'], verify=config['verify_ssl'])
        t = etree.XML(r.content)
        challenge = t.xpath('//Challenge/text()')[0]
        response = '{}-{}'.format(challenge, hashlib.md5('{}-{}'.format(challenge, password).encode('utf-16-le')).hexdigest())
        r = requests.get('{}?username={}&response={}'.format(config['url_login'], config['username'], response), verify=config['verify_ssl'])
        t = etree.XML(r.content)
        return t.xpath('//SID/text()')[0]
    except SSLError:
        ssl_error_exit()


def get_uid(config, sid, mac):
    try:
        payload = {'sid': sid, 'page': 'netDev', 'xhrId': 'all'}
        r = requests.post(config['url_data'], data=payload, verify=config['verify_ssl'])
        devs = json.loads(r.content)
        for dev in devs['data']['passive']:
            if dev['mac'] == mac:
                return dev['UID']
        for dev in devs['data']['active']:
            if dev['mac'] == mac:
                return dev['UID']
        return ''
    except SSLError:
        ssl_error_exit()


def get_version(config, sid):
    try:
        payload = {'sid': sid, 'page': 'overview'}
        r = requests.post(config['url_data'], data=payload, verify=config['verify_ssl'])
        reply = json.loads(r.content)
        return reply['data']['fritzos']['nspver'].split().pop(0)
    except SSLError:
        ssl_error_exit()
    except KeyError:
        return '0.0'


def wake_up(config, sid, uid):
    try:
        payload = {'sid': sid, 'dev': uid, 'oldpage': 'net/edit_device.lua', 'page': 'edit_device', 'btn_wake': ''}
        vers = get_version(config, sid)
        if version.parse(vers) <= version.parse('7.24'):
            payload['page'] += '2'
        r = requests.post(config['url_data'], data=payload, verify=config['verify_ssl'])
        if r.headers.get("content-type").startswith('application/json'):
            reply = json.loads(r.content)
            try:
                if reply['data']['btn_wake'] == 'ok':
                    return True
            except KeyError:
                pass
            return False
        elif '"pid":"netDev"' in r.text:
            return True
        else:
            return False
    except SSLError:
        ssl_error_exit()


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('device', help='device name (from config file) to sent wakeup to', default='default', nargs='?')
    parser.add_argument('--config', '-c', default='wakeup.json', metavar='wakeup.json', help='use specified config file')
    parser.add_argument('--ssl-no-verify', '-k', action='store_true', help='ignore ssl certificate verification')
    args, _ = parser.parse_known_args()

    config = get_config(args.config)
    config['verify_ssl'] = not args.ssl_no_verify

    if args.device in config['devices']:
        target_mac = config['devices'][args.device]
    else:
        error_exit("unknown device {}".format(args.device))

    sid = get_sid(config)
    if sid == SID_NOAUTH:
        error_exit("authentication failed")
    else:
        uid = get_uid(config, sid, target_mac)
        if uid:
            if wake_up(config, sid, uid):
                print("[success] wakeup sent to {}".format(target_mac))
            else:
                error_exit("something went wrong while sending wakeup to {}".format(target_mac))
        else:
            error_exit("unknown mac {}".format(target_mac))
